cat/opencl.cat

```diff
@@ -0,0 +1,128 @@
+(* Synchronisation *)
+(*******************)
+
+let Acq = (ACQ | SC | ACQ_REL) & (R | F)
+let Rel = (REL | SC | ACQ_REL) & (W | F)
+
+(* Fences sequenced before or after *)
+let Fsb = [F]; sb
+let sbF = sb; [F]
+
+(* Release sequence *)
+let rs_prime = wi | (unv; [R & W])
+let rs = mo & rs_prime & ~((mo & ~rs_prime) ; mo)
+
+(* Release-acquire synchronisation *)
+let ra_sw(r) = ((r & [Rel]); Fsb?; [W \ WI]; rs?; r; rf; [R \ WI]; sbF?; ([Acq] & r)) & incl & ~wi
+
+(* Barrier synchronisation *)
+// in OpenCL a barrier results in two fence operations: entry and exit fences:
+// https://registry.khronos.org/OpenCL/specs/3.0-unified/html/OpenCL_API.html#_work_group_functions
+let bar_sw(r) = (r & [EF]); syncbar & ~wi & swg; (r & [XF])
+
+(* Allowed to synchronise on the other region *)
+let scf = (SC * SC) | ((dynamic_tag(GLOBAL) & dynamic_tag(LOCAL)); F*F;(dynamic_tag(GLOBAL) & dynamic_tag(LOCAL)))
+
+(* Global and local synchronises-with *)
+let gsw = ra_sw(dynamic_tag(GLOBAL)) | bar_sw(dynamic_tag(GLOBAL)) | (scf & ra_sw(dynamic_tag(LOCAL)))
+let lsw = ra_sw(dynamic_tag(LOCAL)) | bar_sw(dynamic_tag(LOCAL)) | (scf & ra_sw(dynamic_tag(GLOBAL)))
+
+(******************)
+(* Happens-before *)
+(******************)
+
+(* Global and local happens-before *)
+let ghb = ((dynamic_tag(GLOBAL); sb; dynamic_tag(GLOBAL)) | (dynamic_tag(GLOBAL); IW * ~IW; dynamic_tag(GLOBAL)) | gsw)+
+let lhb = ((dynamic_tag(LOCAL); sb; dynamic_tag(LOCAL)) | (dynamic_tag(LOCAL); IW * ~IW; dynamic_tag(LOCAL)) | lsw)+
+
+irreflexive ghb as O-HbG
+irreflexive lhb as O-HbL
+
+(*************)
+(* Coherence *)
+(*************)
+
+let coh(hb) = (rf^-1)?; mo; rf?; hb
+irreflexive coh(ghb) as O-CohG
+irreflexive coh(lhb) as O-CohL
+
+(************************)
+(* Consistency of reads *)
+(************************)
+
+(* A load can only read from a store that already happened. *)
+irreflexive rf; (ghb | lhb) as O-Rf
+
+(* Visible side effects *)
+let vis(hb) = (W * R) & hb & loc & ~((hb & loc); [W]; hb)
+
+(* A non-atomic load can only read from a store that is visible. *)
+empty (rf; (dynamic_tag(NAL) & dynamic_tag(GLOBAL))) \ vis(ghb) as O-NaRfG
+empty (rf; (dynamic_tag(NAL) & dynamic_tag(LOCAL))) \ vis(lhb) as O-NaRfL
+
+(* Consistency of RMWs *)
+empty rmw & (fre; coe) as atomic
+// The original model was tested with Herd, which treats RMW as a single atomic operation.
+// We have modified the model to handle RMW as a sequence of atomic operations.
+// irreflexive rf | (mo;mo;rf^-1) | (mo;rf) as O-Rmw
+
+(****************************************)
+(* Sequential consistency, simplified,  *)
+(* with scoped SC axioms                *)
+(****************************************)
+
+let scp = Fsb?; (rb | mo | (ghb | lhb)); sbF?
+acyclic (SC*SC) & scp & incl as O-Sscoped
+
+(***************)
+(*    Races    *)
+(***************)
+
+let cnf = ((W * W) | (W * R) | (R * W)) & loc
+
+let dr = cnf & ~(ghb | lhb) & ~(ghb | lhb)^-1 & ~wi & ~incl
+flag ~empty dr as data_race
```

Comment:
There is no notion of unsequenced races in opencl?

Reply:
`ur` appears in herd version, overhauling doesn't have one. Shall we add?

Reply:
Yes please. Otherwise, once you add a unit test to compare both models for races, this might cause problems

Reply:
added